File: src/main/java/com/alphawallet/attestation/eip712/Eip712AttestationUsage.java
package com.alphawallet.attestation.eip712;

import com.alphawallet.attestation.FullProofOfExponent;
import com.alphawallet.attestation.IdentifierAttestation.AttestationType;
import com.alphawallet.attestation.SignedIdentityAttestation;
import com.alphawallet.attestation.UseAttestation;
import com.alphawallet.attestation.core.AttestationCrypto;
import com.alphawallet.attestation.core.ExceptionUtil;
import com.alphawallet.attestation.core.SignatureUtility;
import com.alphawallet.attestation.core.URLUtility;
import com.alphawallet.attestation.core.Verifiable;
import com.alphawallet.attestation.eip712.Eip712AttestationUsageEncoder.AttestationUsageData;
import java.io.IOException;
import java.math.BigInteger;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.math.ec.ECPoint;
import org.tokenscript.eip712.Eip712Issuer;
import org.tokenscript.eip712.Eip712Validator;
import org.tokenscript.eip712.JsonEncodable;

/**
 * Class for asserting and validating that a user who already has an Identity Attestation wishes to
 * use it at a given website. The assertion is linked to a public key, which can be used to validate
 * future future communication from the user, up until a user-approved expiration time.
 *
 * Note that this object leaks the user's Identity Attestation (Ethereum Public key) and
 * the user's identifier to the webserver.
 */
public class Eip712AttestationUsage extends Eip712Validator implements JsonEncodable, Verifiable,
    TokenValidateable {
  private static final Logger logger = LogManager.getLogger(Eip712AttestationUsage.class);

  public static final int PLACEHOLDER_CHAIN_ID = 0;

  private final long maxTokenValidityInMs;
  private final UseAttestation useAttestation;
  private final AttestationUsageData data;
  private final String jsonEncoding;
  private final AsymmetricKeyParameter userPublicKey;

  public Eip712AttestationUsage(String attestorDomain, String identifier, UseAttestation useAttestation, AsymmetricKeyParameter signingKey) {
    this(attestorDomain, Timestamp.DEFAULT_TOKEN_TIME_LIMIT,  PLACEHOLDER_CHAIN_ID, identifier, useAttestation, signingKey);
  }

  public Eip712AttestationUsage(String attestorDomain, long maxTokenValidityInMs, long chainId,
      String identifier, UseAttestation useAttestation, AsymmetricKeyParameter signingKey) {
    super(attestorDomain, new Eip712AttestationUsageEncoder(chainId));
    try {
      this.maxTokenValidityInMs = maxTokenValidityInMs;
      this.useAttestation = useAttestation;
      this.jsonEncoding = makeToken(identifier, useAttestation, signingKey);
      this.userPublicKey = retrieveUserPublicKey(jsonEncoding, AttestationUsageData.class);
      this.data = retrieveUnderlyingObject(jsonEncoding, AttestationUsageData.class);
    } catch (Exception e ) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not encode asn1"));
    }
    constructorCheck();
  }

  public Eip712AttestationUsage(String attestorDomain, AsymmetricKeyParameter attestationIssuerVerificationKey,
      String jsonEncoding) {
    this(attestorDomain, attestationIssuerVerificationKey, Timestamp.DEFAULT_TOKEN_TIME_LIMIT, PLACEHOLDER_CHAIN_ID,
        jsonEncoding);
  }

  public Eip712AttestationUsage(String attestorDomain, AsymmetricKeyParameter attestationIssuerVerificationKey,
      long maxTokenValidityInMs, long chainId, String jsonEncoding) {
    super(attestorDomain, new Eip712AttestationUsageEncoder(chainId));
    try {
      this.maxTokenValidityInMs = maxTokenValidityInMs;
      this.jsonEncoding = jsonEncoding;
      this.userPublicKey = retrieveUserPublicKey(jsonEncoding, AttestationUsageData.class);
      this.data = retrieveUnderlyingObject(jsonEncoding, AttestationUsageData.class);
      this.useAttestation = new UseAttestation(URLUtility.decodeData(data.getPayload()), attestationIssuerVerificationKey);
    } catch (Exception e ) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not decode asn1"));
    }
    constructorCheck();
  }

  void constructorCheck() throws IllegalArgumentException {
    if (!verify()) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not verify object"));
    }
  }

  String makeToken(String identifier, UseAttestation useAttestation,
      AsymmetricKeyParameter signingKey) throws IOException {
    Eip712Issuer issuer = new Eip712Issuer<AttestationUsageData>(signingKey, encoder);
    String encodedUseAttestation = URLUtility.encodeData(useAttestation.getDerEncoding());
    Timestamp now = new Timestamp();
    Timestamp expirationTime = new Timestamp(now.getTime() + maxTokenValidityInMs);
    AttestationUsageData data = new AttestationUsageData(
        encoder.getUsageValue(), identifier, encodedUseAttestation, now, expirationTime);
    return issuer.buildSignedTokenFromJsonObject(data, domain);
  }

  private boolean proofLinking() {
    BigInteger candidateExponent = AttestationCrypto.mapToCurveMultiplier(getType(), getIdentifier());
    ECPoint commitmentPoint = AttestationCrypto.decodePoint(getAttestation().getUnsignedAttestation().getCommitment());
    ECPoint candidateRiddle = commitmentPoint.subtract(AttestationCrypto.G.multiply(candidateExponent));
    if (!candidateRiddle.equals(getPok().getRiddle())) {
      logger.error("Could not validate proof linking to attestation commitment");
      return false;
    }
    return true;
  }

  public String getIdentifier() {
    return data.getIdentifier();
  }

  public AsymmetricKeyParameter getUserPublicKey() {
    return userPublicKey;
  }

  public FullProofOfExponent getPok() {
    return useAttestation.getPok();
  }

  public AttestationType getType() {
    return useAttestation.getType();
  }

  public SignedIdentityAttestation getAttestation() {
    return useAttestation.getAttestation();
  }

  public AsymmetricKeyParameter getSessionPublicKey() {
    return useAttestation.getSessionPublicKey();
  }

  @Override
  public String getJsonEncoding() {
    return jsonEncoding;
  }

  @Override
  public boolean checkTokenValidity() {
    long nonceMinTime = Timestamp.stringTimestampToLong(data.getExpirationTime()) - maxTokenValidityInMs;
    long nonceMaxTime = Timestamp.stringTimestampToLong(data.getExpirationTime());
    if (!useAttestation.checkValidity()) {
      logger.error("Not not validate underlying object");
      return false;
    }
    if (!data.getDescription().equals(encoder.getUsageValue())) {
      logger.error("Description field incorrect");
      return false;
    }
    Timestamp time = new Timestamp(data.getTimestamp());
    time.setValidity(maxTokenValidityInMs);
    if (!time.validateAgainstExpiration(Timestamp.stringTimestampToLong(data.getExpirationTime()))) {
      logger.error("Timestamp not valid");
      return false;
    }
    if (!SignatureUtility.verifyKeyAgainstAddress(
        userPublicKey, useAttestation.getAttestation().getUnsignedAttestation().getAddress())) {
      logger.error("Could not verify signature");
      return false;
    }
    if (!Nonce.validateNonce(useAttestation.getPok().getNonce(),
        (useAttestation.getAttestation().getUnsignedAttestation()).getAddress(), domain, new Timestamp(nonceMinTime), new Timestamp(nonceMaxTime))) {
      logger.error("Nonce validation failed");
      return false;
    }
    if (!proofLinking()) {
      logger.error("Could not verify proof linking");
      return false;
    }
    return true;
  }

  @Override
  public boolean verify() {
    if (!useAttestation.verify()) {
      logger.error("Could not verify underlying UseAttestation object");
      return false;
    }
    return true;
  }
}

File: src/main/javascript/crypto/src/libs/Eip712AttestationUsage.ts
import {JsonEncodable} from "../intefaces/JsonEncodable";
import {Verifiable} from "./Verifiable";
import {KeyPair} from "./KeyPair";
import {SignatureUtility} from "./SignatureUtility";
import {Eip712Token} from "./Eip712Token";
import {UseAttestation} from "./UseAttestation";
import {base64ToUint8array, hexStringToBase64Url, logger, uint8tohex} from "./utils";
import {AttestationCrypto, Pedestren_G} from "./AttestationCrypto";
import {CURVE_BN256, Point} from "./Point";
import {FullProofOfExponent} from "./FullProofOfExponent";
import {Nonce} from "./Nonce";
import {TokenValidateable} from "./TokenValidateable";
import {Timestamp} from "./Timestamp";
import {DEBUGLEVEL} from "../config";

export class Eip712AttestationUsage extends Eip712Token implements JsonEncodable, Verifiable, TokenValidateable {
    public PLACEHOLDER_CHAIN_ID: number = 0;
    public Eip712PrimaryName: string = "AttestationUsage";
    public Eip712Description: string = "Prove that the \"identifier\" is the identifier hidden in attestation contained in\"payload\".";
    public Eip712UserTypes: {name: string, type: string}[]  = [
        {name: 'description', type: 'string'},
        {name: 'identifier', type: 'string'},
        {name: 'payload', type: 'string'},
        {name: 'timestamp', type: 'string'},
        {name: 'expirationTime', type: 'string'},
    ]

    private useAttestation: UseAttestation;
    private jsonEncoding: string;
    private attestorKey: KeyPair;
    private userKey: KeyPair;
    private maxTokenValidityInMs: number;

    protected data: {
        payload: string,
        description: string,
        identifier: string,
        timestamp: string,
        expirationTime: string,
    }

    constructor(userKey: KeyPair = null, maxTokenValidityInMs:number = Timestamp.DEFAULT_TOKEN_TIME_LIMIT) {
        super();
        this.maxTokenValidityInMs = maxTokenValidityInMs;
        this.userKey = userKey;
    }

    // TODO make signingKey universal
    public async addData(attestorDomain: string, identifier: string, useAttestation: UseAttestation) {
        this.setDomain(attestorDomain);
        this.useAttestation = useAttestation;

        try {
            this.jsonEncoding = await this.makeToken(identifier, useAttestation);
        } catch ( e ) {
            logger(DEBUGLEVEL.LOW, e);
            throw new Error("Could not encode object. " + e);
        }

        try {
            // decode JSON and fill publicKey
            this.fillJsonData(this.jsonEncoding);
        } catch (e){
            logger(DEBUGLEVEL.LOW, e);
            return false;
        }

        this.constructorCheck();
    }

    fillJsonData(json: string, attestorKey: KeyPair = null){
        if (!json) {
            throw new Error('Empty json');
        }

        if (attestorKey !== null) {
            this.attestorKey = attestorKey;
        }

        this.jsonEncoding = json;
        let tokenData = JSON.parse(json);
        let signatureInHex = tokenData.signatureInHex;

        let jsonSigned = JSON.parse(tokenData.jsonSigned);
        this.eip712DomainData = jsonSigned.domain;
        this.data = jsonSigned.message;

        try {
            let publicKey = SignatureUtility.recoverPublicKeyFromTypedMessageSignature(jsonSigned, signatureInHex);
            this.requestorKeys = KeyPair.fromPublicHex(publicKey.substr(2));
            logger(DEBUGLEVEL.HIGH, 'restored address: ' + this.requestorKeys.getAddress());
        } catch (e){
            let m = "Recover Address failed with error:" + e;
            logger(DEBUGLEVEL.LOW, m, e);
            throw new Error(m);
        }

        if (!this.useAttestation){
            try {
                logger(DEBUGLEVEL.VERBOSE, uint8tohex(base64ToUint8array(this.data.payload)));
                this.useAttestation = UseAttestation.fromBytes(base64ToUint8array(this.data.payload), this.attestorKey);
            } catch (e){
                let m = "Failed to read UseAttestation. " + e;
                logger(DEBUGLEVEL.LOW, m, e);
                throw new Error(m);
            }
        }

        this.constructorCheck();
    }

    constructorCheck() {
        if (!this.verify()) {
            throw new Error("Could not verify Eip712 use attestation");
        }
    }

    // use Att
    async makeToken(identifier: string, useAttestation: UseAttestation): Promise<string>{
        if (!this.userKey) {
            await SignatureUtility.connectMetamaskAndGetAddress();
        }

        let userData = {
            payload: hexStringToBase64Url(useAttestation.getDerEncoding()),
            description: this.Eip712Description,
            timestamp: new Timestamp().getTimeAsString(),
            identifier: identifier,
            expirationTime: new Timestamp(Date.now() + this.maxTokenValidityInMs).getTimeAsString(),
        };

        return await SignatureUtility.signEIP712WithBrowserWallet(this.domain, userData, this.Eip712UserTypes, this.Eip712PrimaryName, this.userKey );
    }

    proofLinking() {

        let crypto = new AttestationCrypto();
        let candidateExponent = crypto.mapToCurveMultiplier(this.getType(), this.getIdentifier());
        let commitmentPoint: Point = Point.decodeFromUint8(this.getAttestation().getUnsignedAttestation().getCommitment(), CURVE_BN256);
        let candidateRiddle: Point = commitmentPoint.subtract(Pedestren_G.multiplyDA(candidateExponent));

        if (!candidateRiddle.equals(this.getPok().getRiddle())) {
            logger(DEBUGLEVEL.LOW, 'candidateRiddle.equals(this.getPok().getRiddle()) error');
            return false;
        }
        return true;
    }

    getPok(): FullProofOfExponent{
        return this.useAttestation.getPok();
    }

    getType(): number {
        return this.useAttestation.type;
    }

    getIdentifier(): string {
        return this.data.identifier;
    }

    getAttestation(){
        return this.useAttestation.getAttestation();
    }

    getJsonEncoding(): string {
        return this.jsonEncoding;
    }

    checkTokenValidity(): boolean {

        let nonceMinTime: number = Timestamp.stringTimestampToLong(this.data.expirationTime) - this.maxTokenValidityInMs - 2 * Timestamp.ALLOWED_ROUNDING;
        let nonceMaxTime: number = Timestamp.stringTimestampToLong(this.data.expirationTime);

        if (!this.useAttestation.checkValidity()){
            logger(DEBUGLEVEL.LOW, 'useAttestation.checkValidity failed');
            return false;
        };

        if (this.data.description != this.Eip712Description) {
            logger(DEBUGLEVEL.LOW, `wrong description: "${this.data.description}", must be "${this.Eip712Description}"`);
            return false;
        };

        let time: Timestamp = new Timestamp(this.data.timestamp);
        time.setValidity(this.maxTokenValidityInMs);
        if (!time.validateAgainstExpiration(Timestamp.stringTimestampToLong(this.data.expirationTime))) {

            logger(DEBUGLEVEL.LOW, 'verify timestamp failed.\n' + this.data.timestamp + "\n" + this.maxTokenValidityInMs + "\n" + this.data.expirationTime + "\n" + Timestamp.stringTimestampToLong(this.data.expirationTime) + "\n");
            return false;
        }

        if (this.requestorKeys.getAddress().toLowerCase() !== this.useAttestation.getAttestation().getUnsignedAttestation().getAddress().toLowerCase()) {
            logger(DEBUGLEVEL.LOW, 'wrong address');
            return false;
        };

        if (!(new Nonce().validateNonce(
            this.useAttestation.getPok().getNonce(),
            (this.useAttestation.getAttestation().getUnsignedAttestation()).getAddress(),
            this.domain,
            nonceMinTime,
            nonceMaxTime
        ))) {
            logger(DEBUGLEVEL.LOW, 'wrong Nonce');
            return false;
        };

        if (!this.proofLinking()) {
            logger(DEBUGLEVEL.LOW, 'wrong proofLinking');
            return false;
        };

        return true;
    }


    verify(): boolean {
        if (!this.useAttestation.verify()) {
            return false;
        }
        return true;
    }

    public getSessionPublicKey(): KeyPair {
        return this.useAttestation.getSessionPublicKey();
    }

}
